scripts/testing/e2e-runner

```diff
@@ -0,0 +1,396 @@
+             --local "$CLONED_REPO" \
+             --results "$RESULT_ROOT" \
+             --name "$RESULT_NAME" \
+             "$E2E_TESTS"
+    else
+        must-cd "$WORKTREE"
+    fi
+}
+
+prepare_result_dir() {
+    local dir="$RESULT_DIR"
+    local log="$dir/${SELF##*/}.log"
+
+    info "creating result dir $dir..."
+    must-mkdir "$dir" "result"
+
+    if ! touch "$log"; then
+        fatal "failed to create log file $log"
+    fi
+    LOG="$log"
+
+    info "Redirecting output to log file $LOG..."
+    if ! exec &>"$LOG"; then
+        fatal "log redirection failed"
+    fi
+}
+
+run_e2e_tests() {
+    local tests
+
+    info "running e2e tests ${E2E_TESTS:-with the default test set}..."
+    must-cd "$WORKTREE"
+    if [ -n "$E2E_TESTS" ]; then
+        make E2E_TESTS="$E2E_TESTS" e2e-tests
+    else
+        make e2e-tests
+    fi
+}
+
+cleanup_test_vms() {
+    local src
+
+    info "cleaning up test VMs..."
+    must-cd "$WORKTREE"
+    for src in test/e2e/n[0-9]*c[0-9]*-c*; do
+        (cd "$src" && vagrant destroy --force)
+    done
+}
+
+collect_git_info() {
+    local describe="$RESULT_DIR/$GIT_DESC"
+    local sha1="$RESULT_DIR/$GIT_SHA1"
+
+    info "collecting info about worktree..."
+    must-cd "$WORKTREE"
+    git describe > "$describe"
+    git log -1 --pretty=%H > "$sha1"
+}
+
+collect_results() {
+    local src dst
+
+    info "collecting test results..."
+    must-cd "$WORKTREE"
+    for src in test/e2e/n[0-9]*c[0-9]*-c*; do
+        dst=$RESULT_DIR/${src#test/e2e/}
+        must-mkdir "$dst" "result"
+        must-mv "$src/$TEST_SUITE_DIR" "$dst"
+    done
+}
+
+cleanup_worktree() {
+    info "cleaning up worktree $WORKTREE..."
+    must-cd "$WORKTREE"
+    if ! git clean -fdx; then
+        fatal "failed to git clean worktree $WORKTREE"
+    fi
+
+    info "removing worktree $WORKTREE..."
+    must-cd "$CLONED_REPO"
+    if ! git worktree remove "$WORKTREE"; then
+        fatal "failed to remove worktree $WORKTREE..."
+    fi
+}
+
+generate_summary() {
+    local summary="$RESULT_DIR/summary.txt"
+    local vm dir f test result failed
+
+    info "generating summary $summary..."
+    rm -f "$summary"
+    if ! touch "$summary"; then
+        error "failed to create $summary"
+        return
+    fi
+
+    for vm in $RESULT_DIR/n[0-9]*c[0-9]*-c*; do
+        echo "${vm##*/}:" >> "$summary"
+        for dir in $(find $vm/$TEST_SUITE_DIR -name 'test[0-9]*-*' -type d); do
+            for f in  $dir/*.output; do
+                mv $f $f.txt
+            done
+            test="${dir##*$TEST_SUITE_DIR/}"
+            result="- FAIL"
+            if [ -f "$dir/summary.txt" ]; then
+                if grep -q PASS "$dir/summary.txt"; then
+                    result="+ PASS"
+                fi
+            fi
+            if [ "$result" = "FAIL" ]; then
+                failed=true
+            fi
+            echo "  $result $test"
+        done | sort >> "$summary"
+    done
+}
+
+generate_index() {
```

Comment:
Just a thought that shouldn't ideally the lower-level test runner script create this one? No need to change it for this pr...

Reply:
Well, IMO ideally neither this nor the lower-level one should create the index. It should be created by the entity which is serving out the test results over HTTP[S]. But to get us started, I hacked it in now here...

Reply:
But if either one is creating, it should be this one. The lower-lever test runner has nothing to do with HTTP.